src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs

```diff
@@ -1025,6 +1066,223 @@ private static string GetFirstLineSubString(string stringToComplete, out bool ha
+            if (assignmentAst.Left is ConvertExpressionAst convertExpressionAst)
+            {
+                typeConstraint = convertExpressionAst.StaticType;
+
+                if (typeConstraint.IsEnum)
+                {
+                    completions = GetResultForEnum(typeConstraint, completionContext);
+                    return true;
+                }
+            }
+
+            // If the assignment itself was unconstrained, the variable still might be
+            if (!TryGetTypeConstraintOnVariable(completionContext, variableAst.VariablePath.UserPath, out typeConstraint, out setConstraint))
+            {
+                return false;
+            }
+
+            // Again try the [ValidateSet()] constraint first
+            if (setConstraint?.ValidValues != null)
+            {
+                completions = GetResultForSet(typeConstraint, setConstraint.ValidValues, completionContext);
+                return true;
+            }
+
+            // Then try to complete for an enum type
+            if (typeConstraint.IsEnum)
+            {
+                completions = GetResultForEnum(typeConstraint, completionContext);
+                return true;
+            }
+
+            return false;
+        }
+
+        private static List<CompletionResult> GetResultForSet(
+            Type typeConstraint,
+            IList<string> validValues,
+            CompletionContext completionContext)
+        {
+            var allValues = new List<string>();
+            foreach (string value in validValues)
+            {
+                if (typeConstraint != null && (typeConstraint == typeof(string) || typeConstraint.IsEnum))
+                {
+                    allValues.Add(GetQuotedString(value, completionContext));
+                }
+                else
+                {
+                    allValues.Add(value);
+                }
+            }
+
+            return GetMatchedResults(allValues, completionContext);
+        }
+
+        private static List<CompletionResult> GetMatchedResults(
+            List<string> allValues,
+            CompletionContext completionContext)
+        {
+            var stringToComplete = string.Empty;
+            if (completionContext.TokenAtCursor != null)
+            {
+                stringToComplete = completionContext.TokenAtCursor.Text;
+            }
+
+            IEnumerable<string> matchedResults = null;
+
+            if (!string.IsNullOrEmpty(stringToComplete))
+            {
+                string matchString = stringToComplete + "*";
+                var wildcardPattern = WildcardPattern.Get(matchString, WildcardOptions.IgnoreCase | WildcardOptions.CultureInvariant);
+
+                matchedResults = allValues.Where(r => wildcardPattern.IsMatch(r));
+            }
+
+            if (matchedResults == null || !matchedResults.Any())
+            {
+                matchedResults = allValues;
```

Comment:
This is not right. It could be that no result matches the `stringToComplete`, and in that case tab completion shouldn't return anything.
This should be changed to
```
if (string.IsNullOrEmpty(stringToComplete))
{
    matchedResults = allValues;
}
else
{
    string matchString = stringToComplete + "*";
    ...
}
```

Reply:
will fix